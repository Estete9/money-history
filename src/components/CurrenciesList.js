import { useDispatch, useSelector } from 'react-redux';
import { React, useEffect, useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import { Link } from 'react-router-dom';
import SectionHeader from './SectionHeader';
import CurrencyElement from './CurrencyElement';
import styles from '../styles/currenciesList.module.css';
import { fetchSymbolsAPI } from '../redux/currencies/currenciesSlice';

function CurrenciesList() {
  const dispatch = useDispatch();
  const { currenciesData, isLoadingData, error } = useSelector((store) => store.currencies);
  const { searchQuery } = useSelector((store) => store.header);
  const [filteredCurrencies, setFilteredCurrencies] = useState(currenciesData);

  useEffect(() => {
    if (isLoadingData && !currenciesData.length) {
      dispatch(fetchSymbolsAPI());
    }
  }, [currenciesData, dispatch, isLoadingData]);

  useEffect(() => {
    if (currenciesData.length) {
      setFilteredCurrencies(currenciesData);
    }
  }, [currenciesData]);

  // prettier-ignore
  useEffect(() => {
    if (searchQuery && currenciesData.length) {
      const filteredData = currenciesData.filter((currency) => currency.currencySymbol
        .toLowerCase()
        .includes(searchQuery.toLowerCase()));
      setFilteredCurrencies(filteredData);
    }
    if (!searchQuery) {
      setFilteredCurrencies(currenciesData);
    }
  }, [currenciesData, dispatch, searchQuery]);

  if (isLoadingData) {
    return <div>Currencies loading...</div>;
  }

  if (error) {
    return <div>{`We encountered an error: ${JSON.stringify(error)}`}</div>;
  }

  if (!currenciesData.length) {
    return (
      <section>
        <header>
          <h2>There are no currencies for the moment</h2>
          <h4>Our currency list is empty</h4>
        </header>
      </section>
    );
  }

  return (
    <div>
      {/* update this HEADER INFORMATION */}
      <SectionHeader countryName="Euro" value="$1.594" />
      <div className={styles.currenciesGrid}>
        {filteredCurrencies.map((currency) => (
          <Link key={uuidv4()} to={`/currency/${currency.currencySymbol}`}>
            <CurrencyElement currency={currency} />
          </Link>
        ))}
      </div>
    </div>
  );
}

export default CurrenciesList;
