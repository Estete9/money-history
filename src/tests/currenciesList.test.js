import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import CurrenciesList from '../components/CurrenciesList';

const mockStore = configureStore([]);

describe('Integration tests for CurrenciesList component', () => {
  test('Checks if loading screen in CurrencyList renders correctly', () => {
    const currenciesStore = mockStore({
      currencies: {
        currenciesData: [],
        isLoadingData: true,
      },
      header: {
        searchQuery: '',
        isSearchBarOpen: '',
      },
    });
    const component = renderer.create(
      <Provider store={currenciesStore}>
        <CurrenciesList />
      </Provider>,
    );
    const currenciesListTree = component.toJSON();

    expect(currenciesListTree).toMatchSnapshot();
  });
});
