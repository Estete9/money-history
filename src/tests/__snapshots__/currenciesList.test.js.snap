// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests for CurrenciesList component Checks if CurrencyList renders correctly after async API consumption 1`] = `
<div>
  <section
    className="sectionHeader"
  >
    <div
      className="sectionHeaderImgWrapper"
    >
      <img
        alt="currency icon"
        className="sectionHeaderImg"
        src="svg-mock"
      />
    </div>
    <div
      className="sectionHeaderContentWrapper"
    >
      <h1>
        $ 1
      </h1>
      <p>
        United States Dollar
      </p>
      <p>
        $0.9452 euro
      </p>
    </div>
  </section>
  <div
    className="separator"
  >
    CURRENCIES
  </div>
  <div
    className="currenciesGrid"
  >
    <a
      className="currencyItem color1"
      href="/currency/AUD"
      onClick={[Function]}
    >
      <div
        className="td"
      >
        <div
          className="enterArrowWrapper"
        >
          <img
            alt="enter currency button"
            className="enterCurrencyArrow"
            src="svg-mock"
          />
        </div>
        <img
          alt="currencyImg"
          className="currencyElementImg"
        />
        <div
          className="currencyContent"
        >
          <p>
            Australian Dollar
          </p>
          <p>
            AUD
          </p>
        </div>
      </div>
    </a>
    <a
      className="currencyItem color2"
      href="/currency/CAD"
      onClick={[Function]}
    >
      <div
        className="td"
      >
        <div
          className="enterArrowWrapper"
        >
          <img
            alt="enter currency button"
            className="enterCurrencyArrow"
            src="svg-mock"
          />
        </div>
        <img
          alt="currencyImg"
          className="currencyElementImg"
        />
        <div
          className="currencyContent"
        >
          <p>
            Canadian Dollar
          </p>
          <p>
            CAD
          </p>
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`Integration tests for CurrenciesList component Checks if loading screen in CurrencyList renders correctly 1`] = `
<div>
  Currencies loading...
</div>
`;
